src/communicator.cu

```diff
@@ -780,3 +837,94 @@ UCXCommunicator* initialize_ucx_communicator(bool use_buffer_communicator,
         return communicator;
     }
 }
+
+
+void NCCLCommunicator::initialize()
+{
+    Communicator::initialize();
+
+    ncclUniqueId nccl_id;
+    if (mpi_rank == 0)
+        NCCL_CALL( ncclGetUniqueId(&nccl_id) );
+    MPI_CALL( MPI_Bcast(&nccl_id, sizeof(nccl_id), MPI_BYTE, 0, MPI_COMM_WORLD) );
+    NCCL_CALL( ncclCommInitRank(&nccl_comm, mpi_size, nccl_id, mpi_rank) );
+
+    CUDA_RT_CALL( cudaStreamCreate(&comm_stream) );
+}
+
+
+void NCCLCommunicator::start()
+{
+    NCCL_CALL( ncclGroupStart() );
+    comm_buffers.clear();
+    comm_buffer_sizes.clear();
+    recv_buffers.clear();
+    recv_buffer_idx.clear();
+}
+
+
+void NCCLCommunicator::send(const void *buf, int64_t count, int element_size, int dest)
+{
+    // Note: NCCL has performance issue when the buffer is not 128-bit aligned.
+    // This implementation gets away with issue by forcing the communication to go through a
+    // allocated communication buffer. This buffer is 256B aligned to be compatible with RMM.
+    std::size_t aligned_size = (count * element_size + 255) / 256 * 256;
+
+    comm_buffers.push_back(
+        rmm::mr::get_current_device_resource()->allocate(aligned_size, cudaStreamDefault));
+    CUDA_RT_CALL( cudaStreamSynchronize(cudaStreamDefault) );
+    comm_buffer_sizes.push_back(count * element_size);
+
+    CUDA_RT_CALL(
+        cudaMemcpy(comm_buffers.back(), buf, count * element_size, cudaMemcpyDeviceToDevice
```

Comment:
Why this is not `cudaMemcpyAsync(..., comm_stream)`? Is the call synchronous on purpose?

Reply:
NCCL documentation says:
```
Caution: When called inside a group, stream operations (like ncclAllReduce) can return without having enqueued the operation on the stream. Stream operations like cudaStreamSynchronize can therefore be called only after ncclGroupEnd returns.
```

I think this means `cudaMemcpyAsync` on `comm_stream` is not guaranteed to finish before the buffer is sent to remote GPUs.